lib/qa_utils.cc

```diff
@@ -440,164 +440,324 @@ inline void run_cast_test3_s32fc(volk_fn_3arg_s32fc func,
-                fail = true;
-                if (print_max_errs-- > 0) {
-                    std::cout << "offset " << i / 2 << " in1: " << in1[i] << " + "
-                              << in1[i + 1] << "j  in2: " << in2[i] << " + " << in2[i + 1]
-                              << "j";
-                    std::cout << " tolerance was: " << tol << std::endl;
+                // the primary test is the percent different greater than given tol
+                else {
+                    rel_err = err / norm;
+                    if (rel_err > tol) {
+                        this_fail = true;
+                    }
                 }
             }
         }
+        if (rel_err > max_rel_err) {
+            max_rel_err = rel_err;
+        }
+        if (this_fail) {
+            fail = true;
+            fail_indices.push_back(i / 2);
+        }
     }
 
     return fail;
 }
 
 template <class t>
-bool icompare(t* in1, t* in2, unsigned int vlen, unsigned int tol)
+bool icompare(t* expected,
+              t* actual,
+              unsigned int vlen,
+              unsigned int tol,
+              std::vector<unsigned int>& fail_indices,
+              double& max_rel_err)
 {
     bool fail = false;
-    int print_max_errs = 10;
+    max_rel_err = 0.0;
     for (unsigned int i = 0; i < vlen; i++) {
-        if (((uint64_t)abs(int64_t(((t*)(in1))[i]) - int64_t(((t*)(in2))[i]))) > tol) {
+        t exp_val = expected[i];
+        t act_val = actual[i];
+        uint64_t abs_err = (uint64_t)abs(int64_t(exp_val) - int64_t(act_val));
+        double rel_err =
+            (exp_val != 0) ? (double)abs_err / fabs((double)exp_val) : (double)abs_err;
+        if (rel_err > max_rel_err) {
+            max_rel_err = rel_err;
+        }
+        if (abs_err > tol) {
             fail = true;
-            if (print_max_errs-- > 0) {
-                std::cout << "offset " << i
-                          << " in1: " << static_cast<int64_t>(t(((t*)(in1))[i]))
-                          << " in2: " << static_cast<int64_t>(t(((t*)(in2))[i]));
-                std::cout << " tolerance was: " << tol << std::endl;
-            }
+            fail_indices.push_back(i);
         }
     }
 
     return fail;
 }
 
+// Print error table for failed comparisons
+// Shows: index, input(s), expected, actual, rel_error, tol
+void print_error_table(const std::vector<unsigned int>& fail_indices,
+                       const std::vector<void*>& inputs,
+                       const std::vector<volk_type_t>& input_sigs,
+                       void* expected,
+                       void* actual,
+                       const volk_type_t& output_sig,
+                       float tol,
+                       int max_errors = 10)
+{
+    if (fail_indices.empty())
+        return;
+
+    // Column widths
+    const int w_idx = 7;
+    const int w_val = 10;
+    const int w_sci = 9;
+
+    // Print header
+    std::cout << std::setw(w_idx) << "index";
```

Comment:
Since we rely on C++20 for our tests: https://github.com/gnuradio/volk/blob/5fdfbc340e718a2415a8923141f7964120a32fc5/tests/CMakeLists.txt#L29
You could use std::format to make your life easier.

Reply:
fmt::format should be fine.